azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+        return (int) transOperator.getLastInsertId();
+      }
+      return versionSets.get(0).getVersionSetId();
+    };
+    try {
+      Integer versionSetId = this.dbOperator.transaction(insertVersionSet);
+      VersionSet versionSet = new VersionSet(versionSetJsonString, versionSetMd5Hex, versionSetId);
+      this.md5ToVersionSet.put(versionSetMd5Hex, versionSet);
+      this.idToVersionSet.put(versionSetId, versionSet);
+      return versionSetId;
+    } catch (SQLException e) {
+      logger.error("Exception occurred while inserting version set and getting version id", e);
+      throw new IOException("Unable to insert and get versionSetId", e);
+    }
+  }
+
+  /**
+   * Deletes the versionSet row from the table version_set corresponding to versionSetMd5Hex.
+   */
+  @Override
+  public synchronized boolean deleteVersionSet(String versionSetMd5Hex) throws IOException {
+    final SQLTransaction<Integer> insertVersionSet = transOperator -> transOperator
+        .update(DELETE_VSET, versionSetMd5Hex);
+    try {
+      Integer rowsDeleted = this.dbOperator.transaction(insertVersionSet);
+      if (this.md5ToVersionSet.containsKey(versionSetMd5Hex)) {
+        VersionSet removedVersionSet = this.md5ToVersionSet.remove(versionSetMd5Hex);
+        this.idToVersionSet.remove(removedVersionSet.getVersionSetId());
+      }
+      return rowsDeleted != 0;
+    } catch (SQLException e) {
+      logger.error("Failed to execute: " + DELETE_VSET, e);
+      throw new IOException("Exception occurred while removing version set for " + versionSetMd5Hex,
+          e);
+    }
+  }
+
+  /**
+   * This method first checks if the versionSetId already exists in the local copy of the
+   * versionSets, otherwise, it will insert the versionSet into the table and get the resulting
+   * versionSetId.
+   */
+  @Override
+  public synchronized int getVersionSetId(String versionSetMd5Hex, String versionSetJsonString)
+      throws IOException {
+    if (this.md5ToVersionSet.containsKey(versionSetMd5Hex)) {
+      VersionSet versionSet = this.md5ToVersionSet.get(versionSetMd5Hex);
+      return versionSet.getVersionSetId();
+    }
+    return insertAndGetVersionSetId(versionSetMd5Hex, versionSetJsonString);
+  }
+
+  /**
+   * This method first check if the versionSet exists in the local copy of versionSets, otherwise,
+   * it gets the copy from the version_set table and updates the local copy.
+   */
+  @Override
+  public synchronized VersionSet getVersionSet(String versionSetMd5Hex) throws IOException {
+    if (this.md5ToVersionSet.containsKey(versionSetMd5Hex)) {
+      return this.md5ToVersionSet.get(versionSetMd5Hex);
+    }
+    final VersionSetHandler versionSetHandler = new VersionSetHandler();
+    try {
+      List<VersionSet> versionSets = this.dbOperator
+          .query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+      if (versionSets.size() > 1) {
+        throw new IOException(
```

Comment:
ImageMgmtDaoException

Reply:
Since VersionSetLoader interface has the definition to throw IOException, it is kept along the same lines.